GDB: add support for Intel ADSP platforms

lyakh opened the pull request:
This is a respin of #49077. Depends on #89612 and aligns with https://github.com/thesofproject/linux/pull/5411. A SOF PR to follow

lyakh committed:
gdb: fix building without UART support

GDB can now also be built without UART support, e.g. on Xtensa it can
use a memory window backend. Update the Kconfig dependency to avoid
build failures.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

lyakh committed:
intel_adsp: ace: add gdb support

Add gdbstub support to Meteor Lake and a base for other ACE
platforms. Based on the Tiger Lake implementation, but some registers
do differ.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

lyakh committed:
xtensa: gdbstub: (cosmetic) simplify a conditional

Move an assignment performed on both "if" branches as the last
operation to after the "if."

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

lyakh committed:
xtensa: gdbstub: update comments

Update a comment about GDB use of register index and "regno"
parameters. Only some GDB versions use "idx" instead of "regno."

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

dcpleung commented:
It would be a good idea to separate the changes from the first commit into: arch, subsys and SoC. The backend changes should be squashed into one commit.

lyakh committed:
soc: intel_adsp: gather debug window indices in one place

All debug window slot indices should be defined in
adsp_debug_window.h, move them there and use them consistently.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

lyakh committed:
soc: intel_adsp: add a debug window slot descriptor

Add a debug window slot descriptor for the debug window area in the
first page, where descriptors themselves occupy the first few
hundreds of bytes. This area begins 1024 bytes after the beginning of
the debug window until the end of the page and is therefore 3 * 1024
bytes large. Descriptor 15 will be used to describe it and it will be
used for the GDB stub.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

nklayman committed:
gdbstub: make immediate GDB stub entry optional

Add a Kconfig option to be able to postpone entering the GDB stub
until an external trigger.

Signed-off-by: Noah Klayman <noah.klayman@intel.com>
Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

nklayman committed:
xtensa: gdbstub: fix unavailable register reporting

To report unavailable registers to the GDB client, the GDB stub has
to return a string like "xx" in the value string.

Signed-off-by: Noah Klayman <noah.klayman@intel.com>
Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

lyakh committed:
xtensa: gdbstub: support more registers

Stack offset for spilled registers can be larger than 256 bytes, e.g.
on ADSP platforms with additional floating point and SIMD registers.
Register indices can also be larger than 255. Update both types to
support larger values.

Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>

nklayman committed:
intel_adsp: cavs: add gdb support

Uses SRAM debug region to pass GDB messages to host, which is exposed
via PTY by cavstool with -d arg or using the SOF kernel driver via a
debugfs file.

Signed-off-by: Noah Klayman <noah.klayman@intel.com>
[guennadi.liakhovetski@linux.intel.com: update to current "main"]
Signed-off-by: Guennadi Liakhovetski <guennadi.liakhovetski@linux.intel.com>
